[WIP] Container Metadata

huerdong opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/aws/amazon-ecs-agent/blob/master/CONTRIBUTING.md

Please provide the following information:
-->

### Summary
<!-- What does this pull request do? -->
Adds the Container Metadata feature to the agent, exposing the metadata from ECS and from Docker to agent managed containers, such as the container's DockerID and the Container Instance ARN, and then exposes this metadata for users to access from within the container. 

### Implementation details
<!-- How are the changes implemented? -->
See #848 

### Testing
<!-- How was this tested? -->
<!--
Note for external contributors:
`make test` and `make run-integ-tests` can run in a Linux development
environment like your laptop.  `go test -timeout=25s ./agent/...` and
`.\scripts\run-integ.tests.ps1` can run in a Windows development environment
like your laptop.  Please ensure unit and integration tests pass (on at least
one platform) before opening the pull request.  `make run-functional-tests` and
`.\scripts\run-functional-tests.ps1` must be run on an EC2 instance with an
instance profile allowing it access to AWS resources.  Running
`make run-functional-tests` and `.\scripts\run-functional-tests.ps1` may incur
charges to your AWS account; if you're unable or unwilling to run these tests
in your own account, we can run the tests and provide test results.
-->
- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: <!-- yes|no --> Yes

### Description for the changelog
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
You can see our changelog entry style here:
https://github.com/aws/amazon-ecs-agent/commit/c9aefebc2b3007f09468f651f6308136bd7b384f
-->
Feature - Add metadata service for ECS-created containers. 

### Licensing
<!--
Please confirm that this contribution is under the terms of the Apache 2.0
License.
-->
This contribution is under the terms of the Apache 2.0 License: <!-- yes --> Yes

sharanyad commented:
Shouldn't this be merged into dev instead of master?

vsiddharth commented:
@huerdong please change the base branch

huerdong commented:
Changing the base branch seems to be causing major problems in the commit history.

samuelkarp commented:
You'll need to change the base branch before we can review this.  Right now, this pull request shows commits that are already merged into the dev branch from different authors; it's not practical to review that mix of commits.

vsiddharth commented:
@huerdong Please update the summary block and check the appropriate boxes. 
Please double check the license attribution too. 

Thanks

aaithal commented on agent/containermetadata/metadata_manager.go:
```diff
@@ -0,0 +1,144 @@
+
+	err = os.MkdirAll(metadataDirectoryPath, os.ModePerm)
+	if err != nil {
+		return err
```
Please add more context for this error. At the very least something like `fmt.Errorf("creating metadata directory: %v", err)`.

vsiddharth commented on README.md:
```diff
@@ -175,6 +175,8 @@ configure them as something other than the defaults.
 | `ECS_IMAGE_MINIMUM_CLEANUP_AGE` | 30m | The minimum time interval between when an image is pulled and when it can be considered for automated image cleanup. | 1h | 1h |
 | `ECS_NUM_IMAGES_DELETE_PER_CYCLE` | 5 | The maximum number of images to delete in a single automated image cleanup cycle. If set to less than 1, the value is ignored. | 5 | 5 |
 | `ECS_INSTANCE_ATTRIBUTES` | `{"stack": "prod"}` | These attributes take effect only during initial registration. After the agent has joined an ECS cluster, use the PutAttributes API action to add additional attributes. For more information, see [Amazon ECS Container Agent Configuration](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-agent-config.html) in the Amazon ECS Developer ...
+| `ECS_ENABLE_CONTAINER_METADATA` | `<true|false>` | Whether to enable creation of a metadata file in the container | `false` | `false` |
```
Please update this line with more information

vsiddharth commented on README.md:
```diff
@@ -175,6 +175,8 @@ configure them as something other than the defaults.
 | `ECS_NUM_IMAGES_DELETE_PER_CYCLE` | 5 | The maximum number of images to delete in a single automated image cleanup cycle. If set to less than 1, the value is ignored. | 5 | 5 |
 | `ECS_INSTANCE_ATTRIBUTES` | `{"stack": "prod"}` | These attributes take effect only during initial registration. After the agent has joined an ECS cluster, use the PutAttributes API action to add additional attributes. For more information, see [Amazon ECS Container Agent Configuration](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-agent-config.html) in the Amazon ECS Developer ...
+| `ECS_ENABLE_CONTAINER_METADATA` | `<true|false>` | Whether to enable creation of a metadata file in the container | `false` | `false` |
+| `ECS_HOST_DATA_DIR` | `/var/lib/ecs` | The source directory on the host from which ECS_DATADIR is mounted | `/var/lib/ecs` | `Not used` |
```
Update this line too

sharanyad replied:
Could you please add information on why this is not used for Windows?

aaithal commented on agent/containermetadata/metadata_manager.go:
```diff
@@ -0,0 +1,144 @@
+	// Create metadata file
+	err = createMetadataFile(metadataDirectoryPath)
+	if err != nil {
+		return err
```
Same as above. This error needs context. You can even use `errors.Wrapf()` from https://github.com/pkg/errors

aaithal commented on agent/containermetadata/parse_metadata.go:
```diff
@@ -0,0 +1,153 @@
+// configuration and data then packages it for JSON Marshaling
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this or the above stage.
+func (manager *metadataManager) parseMetadata(dockerContainer *docker.Container, task *api.Task, container *api.Container) Metadata {
```
It seems like all you need here are task's arn and container's name. Can you just pass those instead of `task` and `container`?

aaithal commented on agent/containermetadata/parse_metadata.go:
```diff
@@ -0,0 +1,153 @@
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this or the above stage.
+func (manager *metadataManager) parseMetadata(dockerContainer *docker.Container, task *api.Task, container *api.Container) Metadata {
+	taskMD := parseTaskMetadata(task, container)
```
Along the same lines, I don't see the value of creating a function which just initializes a struct. It would be lot helpful to anyone who reads the code to just do this instead:
` return Metadata { taskMetadata: TaskMetadata{...} ...}`

aaithal commented on agent/containermetadata/types.go:
```diff
@@ -0,0 +1,109 @@
+type Network struct {
+	NetworkMode string `json:"NetworkMode,omitempty"`
+	IPv4Address string `json:"IPv4Address,omitempty"`
+}
```
Please add the `IPv6Address` field here as well with `omitempty`. You don't need to do anything for it right now. But, it'll be useful when "Task ENIs" is enabled.

vsiddharth commented on agent/containermetadata/metadata_manager.go:
```diff
@@ -0,0 +1,144 @@
+
+// metadataManager implements the MetadataManager interface
+type metadataManager struct {
+	client               dockerMetadataClient
```
I'd like to see some comments here for each attribute

aaithal commented on agent/containermetadata/write_metadata.go:
```diff
@@ -0,0 +1,64 @@
+// files of a given task
+func getTaskMetadataDir(task *api.Task, dataDir string) (string, error) {
+	taskID, err := getTaskIDfromARN(task.Arn)
+	return filepath.Join(dataDir, metadataJoinSuffix, taskID), err
```
You need to deal with `err` here.

aaithal commented on agent/containermetadata/write_metadata_unix.go:
```diff
@@ -0,0 +1,84 @@
+		return err
+	}
+	defer temp.Close()
+	if err != nil {
```
This is redundant

aaithal commented on agent/containermetadata/write_metadata.go:
```diff
@@ -0,0 +1,64 @@
+
+const (
+	metadataJoinSuffix = "metadata"
+	metadataFile       = "metadata.json"
```
How about "ecs-container-metadata.json" ?

aaithal commented on agent/engine/docker_task_engine.go:
```diff
@@ -303,6 +308,14 @@ func (engine *DockerTaskEngine) sweepTask(task *api.Task) {
+	if engine.cfg.ContainerMetadataEnabled {
+		err := engine.metadataManager.CleanTaskMetadata(task)
+		if err != nil {
+			seelog.Errorf("clean task metadata failed for task %s: %v", task, err)
```
Logs need to start with Upper case letters.

vsiddharth commented on agent/containermetadata/parse_metadata.go:
```diff
@@ -0,0 +1,153 @@
+		cluster:              manager.cluster,
+		taskMetadata:         taskMD,
+		containerInstanceARN: manager.containerInstanceARN,
+		metadataStatus:       "NOT_READY",
```
Could we have a new type for metadata status ?
I'd like to have it be one amongst a set of well defined values

vsiddharth commented on agent/containermetadata/write_metadata_unix.go:
```diff
@@ -0,0 +1,84 @@
+	if err != nil {
+		return fmt.Errorf("write to metadata file for task %s container %s: %v", task, container, err)
+	}
+	metadataFilePath := filepath.Join(metadataFileDir, metadataFile)
```
This seems to be duplicated code from the `createMetadataFile` method

vsiddharth commented on agent/containermetadata/write_metadata_windows.go:
```diff
@@ -0,0 +1,79 @@
+	ContainerMetadataClientVersion = dockerclient.Version_1_24
+)
+
+// createMetadataFile initializes the metadata file
```
Update this comment

vsiddharth commented on agent/containermetadata/write_metadata_windows.go:
```diff
@@ -0,0 +1,79 @@
+func createMetadataFile(metadataDirectoryPath string) error {
+	metadataFilePath := filepath.Join(metadataDirectoryPath, metadataFile)
+	file, err := os.Create(metadataFilePath)
+	if err != nil {
```
Why didn't you go with the "Tempfile" approach here ?

aaithal commented on agent/app/agent.go:
```diff
@@ -108,6 +110,26 @@ func newAgent(
 		return nil, err
 	}
 
+	var metadataManager containermetadata.MetadataManager
```
`containermetadata.Manager` stutters less than  `containermetadata.MetadataManager`

huerdong replied:
I understand your purpose but I am reluctant to make this change because there are so many structs in the agent with the word Manager in them. Having a single struct called Manager, even if preceded by package containermetadata, feels confusing to me.

aaithal replied:
That's fine. As long as variables are named correctly, it shouldn't matter what the struct name is.

aaithal commented on agent/app/agent.go:
```diff
@@ -108,6 +110,26 @@ func newAgent(
 
+	var metadataManager containermetadata.MetadataManager
+	if cfg.ContainerMetadataEnabled {
+		// Get the oldest Docker Client version with up to date inspect API
```
Please move this to a method. Also, this needs a test in `agent/app/agent_test.go`

huerdong replied:
Right now I am cautious about making changes here because I have encountered in the past import cycles and/or "interface is not implemented" errors when dealing with this section of code. I will see what I can do here.

aaithal replied:
I don't see how moving this to a method would result in import cycles. On the other hand, moving this to a method improves readability and makes it more testable.

huerdong replied:
Do you mean moving it into a method but still within the app package? Maybe I misconstrued your request.

aaithal replied:
> Do you mean moving it into a method but still within the app package?

Yeah.

aaithal commented on agent/config/config.go:
```diff
@@ -315,6 +315,9 @@ func environmentConfig() (Config, error) {
 		seelog.Debugf("Setting instance attribute %v: %v", attributeKey, attributeValue)
 	}
 
+	containerMetadataEnabled := utils.ParseBool(os.Getenv("ECS_ENABLE_CONTAINER_METADATA"), false)
```
needs a test

huerdong replied:
For this and the "needs a test" comments below for the config package, I have added tests to check for my new configuration variables.

aaithal commented on agent/config/config_unix.go:
```diff
@@ -28,6 +28,7 @@ func DefaultConfig() Config {
 		ReservedPorts:               []uint16{SSHPort, DockerReservedPort, DockerReservedSSLPort, AgentIntrospectionPort, AgentCredentialsPort},
 		ReservedPortsUDP:            []uint16{},
 		DataDir:                     "/data/",
+		DataDirOnHost:               "/var/lib/ecs",
```
needs a test

huerdong replied:
see above

aaithal commented on agent/config/config_windows.go:
```diff
@@ -44,6 +44,7 @@ const (
 func DefaultConfig() Config {
 	programData := utils.DefaultIfBlank(os.Getenv("ProgramData"), `C:\ProgramData`)
 	ecsRoot := filepath.Join(programData, "Amazon", "ECS")
+	dataDir := filepath.Join(ecsRoot, "data")
```
needs a test

huerdong replied:
see above

aaithal commented on agent/containermetadata/metadata_manager.go:
```diff
@@ -0,0 +1,151 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
This can be renamed as `manager.go`

huerdong replied:
If I do not rename MetadataManager, is it still appropriate to rename this file? On one hand, there is no confusion with renaming as this file will be isolated in its own package. On the other this breaks the convention of all the other "_manager" files in the code base.

aaithal replied:
That's not a convention, just a sign of us being immature golang coders when that code was written. I don't see a strong argument against renaming this. Please rename both as per my earlier comment.

aaithal commented on agent/containermetadata/metadata_manager.go:
```diff
@@ -0,0 +1,151 @@
+type MetadataStatus string
+
+const (
+	Metadata_NOT_READY = "NOT_READY"
```
lint: `MetadataNotReady` and `MetadataReady`.

vsiddharth replied:
`NOT_READY` is not entirely correct here as we have some data which we consider as as metadata when this status is used. 

I'd prefer statuses like 
* `CREATED`
* `UPDATED`

and documenting what each of them implies

aaithal replied:
If there's a second update, `CREATED` and `UPDATED` are not sufficient to convey that intent. `NOT_READY` indicates helps with this regard.

vsiddharth replied:
I still don't clearly understand the reason behind the second update. 
However, in this case we could then switch to these states:
* `INITIAL`
* `READY`

I'm not in favor of `NOT_READY` as a status here

petderek replied:
>> Why do we need INITIAL, READY, NOT_READY, etc? As a user, do I care about the state of the metadata?

> @petderek it'd have been better if you had just added to the threaded discussion than a comment here. Regardless, as a user, I'd care as it'd help me distinguish between "will i ever get this data" and "i'll eventually get this data if i keep polling"



Without split updates:
[ does data exist? ] => [ does the field I want exist? ] => [Yay, data!]

With:
[ does data exist? ] => [ is the 'ready' field set? ] => [ does the field i want exist? ] => [Yay, data!]

I'm not convinced that splitting up updates is simplifying. Either way, you will still need to poll for metadata. However, now anyone calling this needs to be aware of a different field that guarantees completeness.

aaithal replied:
Because the update is inherently racy, you need a way to determine if the file is in a consumable format. Data not being present and not being updated are two different things and this is what the field helps with. Alternatively, we only create the file when we have all of the data and get rid of the state stuff.

huerdong replied:
The most general way to do this is to have some kind of update counter, but that information feels too abstract and not informative. I think the best course of action here is to agree on what to name the current two states and then when we need additional layers of updates, make changes to these names.

petderek replied:
>  Alternatively, we only create the file when we have all of the data and get rid of the state stuff.

That's what I'd suggest. I think it would simplify our code and make it easier to consume.

huerdong replied:
The issue with getting rid of the state stuff is that we are getting rid a way to detect the metadata race condition where the user wants this data but it is not ready. The container has to start before we get all the data, so it needs a way to check from within itself that the metadata is not ready. This means that the status of the metadata file itself is a container metadata for which we need to provide lookup.

petderek replied:
> The container has to start before we get all the data, so it needs a way to check from within itself that the metadata is not ready.

What makes this a need? 

I don't think we 'need' any of this state stuff. We can just guarantee that metadata is available once the container is running. Adding metadata state introduces complexity that warrants further changes ( more container states, etc).

My concern is that we have started trying to solve a problem that doesn't need to exist in the first place.

huerdong replied:
I understand what you are saying and certainly we do not need the states stuff if we perform a single metadata file write after the container starts and we have all the data. The problem is for users to check that this file is ready in this case, they would either need a delay after container start or a file exist check after the container starts to ensure the file is valid. This puts unnecessary burden on the user to address the race condition when we can simply provide the availability of the metadata as a metadata field as well.

petderek replied:
>  unnecessary burden on the user to address the race condition

What is the race condition?

aaithal commented on agent/containermetadata/metadata_manager.go:
```diff
@@ -0,0 +1,151 @@
+	inspectContainerTimeout     = 30 * time.Second
+)
+
+type MetadataStatus string
```
lint: needs `// MetadataStatus is ...` or `// MetadataStatus specifies..` etc

aaithal commented on agent/containermetadata/metadata_manager.go:
```diff
@@ -0,0 +1,151 @@
+// operations
+type MetadataManager interface {
+	SetContainerInstanceARN(string)
+	CreateMetadata(*docker.Config, *docker.HostConfig, *api.Task, *api.Container) error
```
As per https://github.com/aws/amazon-ecs-agent/pull/913#discussion_r131703627, you should only require task arn and container name in here and all the way down.

aaithal replied:
You can name this as `Create` instead.

aaithal commented on agent/containermetadata/metadata_manager.go:
```diff
@@ -0,0 +1,151 @@
+type MetadataManager interface {
+	SetContainerInstanceARN(string)
+	CreateMetadata(*docker.Config, *docker.HostConfig, *api.Task, *api.Container) error
+	UpdateMetadata(string, *api.Task, *api.Container) error
```
Same as above. As per https://github.com/aws/amazon-ecs-agent/pull/913#discussion_r131703627, you should only require task arn and container name in here and all the way down.

aaithal replied:
You can name this as `Update` instead.

aaithal commented on agent/containermetadata/metadata_manager.go:
```diff
@@ -0,0 +1,151 @@
+
+	// Acquire the metadata then write it in JSON format to the file
+	metadata := manager.parseMetadataAtContainerCreate(task.Arn, container.Name)
+	data, err := json.MarshalIndent(metadata, "", "\t")
```
You need to handle this `err` as well (points to a lacking unit test).

huerdong replied:
While I have handled this error, I'm not sure if it is possible for an error to occur here and whether it is worthwhile to test for it. Since all my fields in the Metadata struct are strings, there should never be an error in the Marshaling. https://stackoverflow.com/questions/33903552/what-input-will-cause-golangs-json-marshal-to-return-an-error

aaithal commented on agent/containermetadata/metadata_manager.go:
```diff
@@ -0,0 +1,151 @@
+
+	// Acquire the metadata then write it in JSON format to the file
+	metadata := manager.parseMetadata(dockerContainer, task.Arn, container.Name)
+	data, err := json.MarshalIndent(metadata, "", "\t")
```
You need to handle this `err` as well (points to a lacking unit test).

huerdong replied:
See above.

aaithal commented on agent/containermetadata/types.go:
```diff
@@ -0,0 +1,110 @@
+
+// MetadataSerializer is an intermediate struct that converts the information
+// in Metadata into information to encode into JSOn
+type MetadataSerializer struct {
```
Can this not be global scoped (`metadataSerializer `)?

aaithal commented on agent/containermetadata/write_metadata.go:
```diff
@@ -0,0 +1,64 @@
+// getTaskMetadataDir acquires the directory with all of the metadata
+// files of a given task
+func getTaskMetadataDir(task *api.Task, dataDir string) (string, error) {
+	taskID, err := getTaskIDfromARN(task.Arn)
```
You need handle this error (points to a lack of unit test).

aaithal commented on agent/containermetadata/metadata_manager.go:
```diff
@@ -0,0 +1,151 @@
+	SetContainerInstanceARN(string)
+	CreateMetadata(*docker.Config, *docker.HostConfig, *api.Task, *api.Container) error
+	UpdateMetadata(string, *api.Task, *api.Container) error
+	CleanTaskMetadata(*api.Task) error
```
You can name this as `Clean` instead.

aaithal commented:
@huerdong can you also paste the output of `go test -cover ./agent/containermetadata`? You may need to make some modifications with the command. I'm mostly looking for test coverage data.

vsiddharth commented:
@huerdong Could you also resolve the conflicts when you update #913 ?

samuelkarp commented on agent/app/agent.go:
```diff
@@ -108,6 +110,26 @@ func newAgent(
+		// Get the oldest Docker Client version with up to date inspect API
+		// as some metadata is unavailable in older API client versions
+		// If this version is unavailable we use default
+		// Due to limitations of interfaces and import cycles
```
I think your only import cycle issue is with the `containermetadata.ContainerMetadataClientVersion`.  I'd be in favor of moving this into the client/factory code directly and just moving the constant over there as well so you don't have an import cycle; that'll keep the impact localized to just the code that needs to be refactored anyway.

huerdong replied:
I'm not quite sure what you mean here. The problem I am encountering is that adding the WithVersion function to the internal dockerMetadataClient interface in the containermetadata package led to "interface not implemented" errors in the past. I am very uncertain about this section of code as a result, but I will try to see if I can minimize the exposure in this section.

samuelkarp commented on agent/app/agent.go:
```diff
@@ -108,6 +110,26 @@ func newAgent(
+	if cfg.ContainerMetadataEnabled {
+		// Get the oldest Docker Client version with up to date inspect API
+		// as some metadata is unavailable in older API client versions
+		// If this version is unavailable we use default
```
It'd be reasonable to just disable the feature if Docker is too old.

huerdong replied:
The feature still works for older versions of docker. The differences in the versions do not actually matter previously because I get the relevant networking metadata through an alternative route for older versions. With the ENI feature added, this difference is significant because older versions of Docker API won't display the additional network modes. Again, I am not very familiar with ENI project except that I was asked to make the Network information into a list to support multiple network modes. I would think if a user wants to use this feature even with an older version of Docker, we should not disable it if it works for that older client.

samuelkarp commented on agent/containermetadata/parse_metadata.go:
```diff
@@ -0,0 +1,136 @@
+// configuration and data then packages it for JSON Marshaling
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this or the above stage.
+func (manager *metadataManager) parseMetadata(dockerContainer *docker.Container, taskARN string, containerName string) Metadata {
```
Generally, I'd like to see the functions in a file organized with caller-above-callee so that you start with abstractions at a high level and get deeper into details as you read further in the file.  That would mean that the entrypoints (`parseMetadata` and `parseMetadataAtContainerCreate`) are at the top, then `parseDockerContainerMetadata`, and last `parseNetworkMetadata`.

huerdong replied:
I have made this reorganization (I'll also use this to remind myself to reorganize the Ports in the metadata which was not addressed yet from a previous comment).

samuelkarp commented on agent/containermetadata/write_metadata_windows.go:
```diff
@@ -0,0 +1,68 @@
+
+	file, err := os.OpenFile(metadataFileName, os.O_WRONLY, 644)
+	if err != nil {
+		// Retry if file does not exist
```
It's worth checking what this `err` is before blindly retrying.

samuelkarp commented on agent/engine/docker_task_engine.go:
```diff
@@ -625,7 +647,21 @@ func (engine *DockerTaskEngine) startContainer(task *api.Task, container *api.Co
+	dockerContainerMD := client.StartContainer(dockerContainer.DockerID, startContainerTimeout)
+
+	// Get metadata through container inspection and available task information then write this to the metadata file
+	// Performs this in the background to avoid delaying container start
```
It almost feels like this should be a new container state.  As is, if the agent restarts while this goroutine is running (after the container has started and we've recorded that in the state file, but before we've completed updating the metadata), you may never get metadata properly updated (`startContainer` won't be called again).

huerdong replied:
I think I previously discussed making this a separate container state with @aaithal. Given the timeline of my ability to work on this project, I do not think I can feasibly make this change right now.

richardpen replied:
This could definite happen in the agent, to make it simple than adding a new state is to add a new state in the container struct to indicate the metadata state.

huerdong replied:
I have added a TODO here for adding a metadata file status in api.Container and for doing metadata updates for containers running after agent restart.

sharanyad commented on agent/containermetadata/write_metadata.go:
```diff
@@ -0,0 +1,65 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Why is this file named 'write_metadata.go'? This does not really write metadata as such.

huerdong replied:
Most of my writing operations was originally here and has since been moved to other places. It is probably better to name it utils.go

sharanyad commented:
Having some unit tests for metadata manager would be good.

petderek commented:
Why do we need INITIAL, READY, NOT_READY, etc? As a user, do I care about the state of the metadata?

aaithal commented:
@petderek  it'd have been easier if you had just added to the threaded discussion than a comment here. Regardless, as a user, I'd care as it'd help me distinguish between "will i ever get this data" and "i'll eventually get this data if i keep polling"

richardpen commented on agent/engine/docker_task_engine.go:
```diff
@@ -597,6 +610,15 @@ func (engine *DockerTaskEngine) createContainer(task *api.Task, container *api.C
 	seelog.Infof("Created container name mapping for task %s - %s -> %s", task, container, containerName)
 	engine.saver.ForceSave()
 
+	// Create metadata directory and file then populate it with common metadata of all containers of this task
```
Looks like the changes in this file are not unit tested.

huerdong replied:
I have added unit tests for the docker_task_engine_test.go for the cases where metadata service is enabled.

richardpen commented on agent/engine/docker_task_engine.go:
```diff
@@ -303,6 +308,14 @@ func (engine *DockerTaskEngine) sweepTask(task *api.Task) {
 	}
+
+	// Clean metadata directory for task
+	if engine.cfg.ContainerMetadataEnabled {
```
This check seems weak, as writing the metadata file could fail. The safer way is to add a new state of this file and only delete this file if it's created.

Also please add a debug log here to indicate this happened.

huerdong replied:
I am not sure why an additional check is needed. If the file is not created, the cleanup would simply do nothing (Not even return an error)
https://golang.org/pkg/os/#RemoveAll

richardpen commented on agent/containermetadata/write_metadata_windows.go:
```diff
@@ -0,0 +1,68 @@
+	}
+	metadataFileName := filepath.Join(metadataFileDir, metadataFile)
+
+	file, err := os.OpenFile(metadataFileName, os.O_WRONLY, 644)
```
can you make `644` a constant, probably move the one defined in `write_metadata_unix.go` to `metadatamanager.go`.

huerdong commented on agent/app/agent.go:
```diff
@@ -108,6 +110,26 @@ func newAgent(
+				break
+			}
+		}
+		metadataManager = containermetadata.NewMetadataManager(metadataDockerClient, cfg)
```
The only missing metadata is in the network settings with the old client. However because we have specific checks in the parsing of the networking metadata to address this issue we will not actually be missing any information unless there are multiple network modes present.

samuelkarp replied:
If we handle old clients, do we even care about the client version anymore?  Should we just use the default client and get rid of this logic entirely?

huerdong replied:
The client version matters because with Task ENI multiple network modes are now supported. The default client does not give information on additional network modes which is only available in v1.21.

samuelkarp replied:
@richardpen and @aaithal can correct me here, but I don't believe that `docker inspect` will return information about the network interfaces added through a CNI plugin.

aaithal replied:
> but I don't believe that docker inspect will return information about the network interfaces added through a CNI plugin

That's correct. It's also worth nothing that we invoke 2 top level CNI plugins and each of them results in a network interface being added to the container (in effect, being part of multiple "network"s)

huerdong replied:
In that case, I will remove the 1.21 client check and use the default. I will leave the Networks field in the metadata stored as a list in case we use a newer API version as a default later.

huerdong commented:
I have added unit tests for the containermetadata package, addressing comments by @sharanyad @aaithal. 
In addition as requested by @aaithal here is the output of `go test -cover ./agent/containermetadata`

`ok  	github.com/aws/amazon-ecs-agent/agent/containermetadata	0.012s	coverage: 74.5% of statements
`

aaithal commented:
@huerdong `74.5%` is a relatively low number for a new package that's being written. Please take another look at what code paths are not being covered in tests and add coverage for the same. Here's a blog on how you can generate html reports to visualize the missing test coverage in code: https://blog.golang.org/cover

opalelement commented:
@aaithal Looks like @huerdong mentioned the code coverage has been bumped to 96.2%

vsiddharth commented:
@opalelement The code coverage was bumped by @huerdong after we introduced him to using wrappers for mocks.

mariuszluciow commented:
Hi here, any news on this? We are waiting for #198 quite some time now...

aaithal commented:
closing this in favor of https://github.com/aws/amazon-ecs-agent/pull/981
